### ISSUE TITLE
Opaque functions to be used in expressions

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#233]
When ks-opaque-types is true also allow opaque functions in expressions.

**Comment by GreyCat** [comment]
We don't have *any* functions in expressions yet.

**Comment by pi-guy-in-the-sky** [comment]
One possible use case: the size of an array is specified using an opaque "variable int" type, and I'd like to be able to `repeat-expr: n_elements`, where `n_elements` is of our variable int type. Is there a way to do this at the moment?

**Comment by KOLANICH** [comment]
Of course there is! Just expose the value of your own type as anything looking like an instance.

**Comment by dgelessus** [comment]
Depending on how complex your variable int type is, you may be able to do this using [value instances](https://doc.kaitai.io/user_guide.html#_value_instances). For some examples, see [`vlq_base128_be`](https://github.com/kaitai-io/kaitai_struct_formats/blob/fa70a3e1baf90517db1145d3c552430bc8d8ad68/common/vlq_base128_be.ksy#L45), [`bcd`](https://github.com/kaitai-io/kaitai_struct_formats/blob/fa70a3e1baf90517db1145d3c552430bc8d8ad68/common/bcd.ksy#L57), or [`dcmp_variable_length_integer`](https://github.com/kaitai-io/kaitai_struct_formats/blob/fa70a3e1baf90517db1145d3c552430bc8d8ad68/macos/resource_compression/dcmp_variable_length_integer.ksy#L62).

When you say "opaque type", do you mean an actual Kaitai Struct [opaque type](https://doc.kaitai.io/user_guide.html#opaque-types) (i. e. structure type defined manually in code instead of in a KSY), or do you just mean that the variable int's internal structure shouldn't be visible to other parts of your KSY? The latter can be done using value instances as explained above. But if you need to use a real opaque type, this isn't possible. As far as I know, there's no way for an opaque type to expose any information to other parts of the KSY.

**Comment by pi-guy-in-the-sky** [comment]
Yes, I was talking about opaque types defined outside of the `.ksy`. I actually found a workaround to access properties of the opaque type using casting.

In my C++ definition of the opaque type, I have:

```cpp
// vint.h

#include <kaitai/kaitaistream.h>

class vint_t
{
    long long val_;

public:
    vint_t(kaitai::kstream *ks);
    long long val();
};
```
```cpp
// vint.cpp

long long vint_t::val()
{
    return val_;
}
```

And then in the `.ksy` file I can access `vint::val()` via:

```
types:
  path:
    seq:
      - id: length
        type: vint
      - id: value
        size: length.val.as<u4>
```

Which generates this code:

```cpp
void my_type_t::path_t::_read() {
    m_length = new vint_t(m__io);
    m_value = m__io->read_bytes(static_cast<uint32_t>(length()->val()));
}
```

Maybe update docs to reflect this?
